src/serial/serial.c

```diff
@@ -94,52 +99,124 @@ struct Serial* serial_create(const char *name, const size_t tx_cap,
         return s;
 }
 
-static void _log(const struct Serial *s, const char *action,
-                 bool *pfx, const char data)
+static void log_header_if_necessary(struct Serial *s,
+                                    const enum data_dir dir)
 {
-        if (!s->log)
+        size_t count = 0;
+        switch (dir) {
+        case DATA_DIR_RX:
+                count = s->log_rx_cntr;
+                break;
+        case DATA_DIR_TX:
+                count = s->log_tx_cntr;
+                break;
+        }
+
+        /* If not at 0'th position, do not print header */
+        if (count)
                 return;
 
-        if (*pfx) {
-                if (s->name) {
-                        pr_info("[serial (");
-                        pr_info(s->name);
-                        pr_info(")] ");
-                } else {
-                        pr_info("[serial] ");
-                }
-                pr_info(action);
-                *pfx = false;
+        pr_info("\r\n");
+        if (s->name) {
+                pr_info("[serial (");
+                pr_info(s->name);
+                pr_info(")] ");
+        } else {
+                pr_info("[serial] ");
         }
 
+        switch (dir) {
+        case DATA_DIR_RX:
+                pr_info("RX: ");
+                s->log_tx_cntr = 0;
+                break;
+        case DATA_DIR_TX:
+                pr_info("TX: ");
+                s->log_rx_cntr = 0;
+                break;
+        }
+}
+
+static void log_ascii(size_t* cntr, const char data)
+{
         switch(data) {
         case('\r'):
                 pr_info("(\\r)");
                 break;
         case('\n'):
-                pr_info("(\\n)\r\n");
-                *pfx = true;
+                pr_info("(\\n)");
+                *cntr = 0;
                 break;
         default:
                 pr_info_char(data);
                 break;
         }
 }
 
+static void log_binary(size_t* cntr, const char data)
+{
+        char buf[12];
+        /* Only print printable characters in data column */
+        if (32 <= data && data <= 126) {
```

Comment:
suggest just printing hex dump of everything if in binary mode. it's unlikely we'll want to see any human readable ascii here